src/upgrades/v02_v03.md

```diff
@@ -0,0 +1,137 @@
+
+```rust
+Range::between(v1, v2.bump())
+```
+
+In v0.3, you can now leverage the official [`core::ops::RangeBounds`](https://doc.rust-lang.org/core/ops/trait.RangeBounds.html) trait.
+
+```rust
+BoundedRange::from_range_bounds(v1..=v2)
+```
+
+To update the usage of the `Range` type in your code should be quite straightforward.
+There has been few naming changes for consistency, such as `Range::none()` becoming `BoundedRange::empty()` but every constructor previously available is still available now.
+Of course, if for any reason the provided `BoundedRange` type does not fit your constraints for handling sets of versions, you can provide your own implementation of the `VersionSet` trait thanks to the new v0.3 API.
+If you are especially interested in v0.3 for dealing with pre-releases, we strongly suggest you also read the [dedicated section of the guide for pre-releases](/limitations/prerelease_versions.md).
+
+## Changes in the DependencyProvider to choose package versions
+
+One fundamental piece of flexibility of the pubgrub crate is that it delegates all decision making to the caller in the `DependencyProvider` trait.
+
+```rust
+/// v0.2 dependency provider trait
+pub trait DependencyProvider<P: Package, V: Version> {
+    /// Given candidates, choose the next package version attempt
+    /// to continue building the solution.
+    fn choose_package_version(&self, potential_packages) -> // (package, version)
+
+    /// Retrieve the dependencies of the given package version.
+    fn get_dependencies(&self, package, version) -> // dependencies
+}
+```
+
+Previously, when choosing the next package to be added to the partial solution being built, the caller would be given a list of candidates, and it has the responsability to pick one package and version satisfying the provided candidates.
+That list of candidates typically grows everytime we encounter a new package and add its dependencies to candidates for the next choices.
+So depending on the resolution path chosen, it is possible to build a list of candidates that grows fast.
+Then, every round, the caller has to work again on the whole list of candidates to pick one package version.
+There is a smarter way to tackle this problem, consisting in puting candidates in a priority queue.
+Everytime the solver encounters a new candidate, it will ask the caller to rate the priority to give for the choice of that candidate.
+Then everytime there is a choice to be made, the algorithm picks the first candidate in the priority list, and simply asks the caller to pick one version in that version set.
+
+```rust
+/// v0.3 dependency provider trait
+pub trait DependencyProvider<P: Package, V: Version> {
+    /// Compute the priority of a given package and range of versions for the solver.
+    /// When deciding which package to pick next, the solver will choose the highest priority one.
+    fn prioritize(&self, package, version_set) -> // priority
```

Comment:
It could help to include an example of the types that might be commonly used for `Priority`.